diff --git a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
index 31b140d..c9b2b2f 100644
--- a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
+++ b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
@@ -132,17 +132,16 @@ async function netlifyAdapter(compilation) {
       projectDirectory,
     );
 
-    redirects += `${route} /.netlify/functions/${id} 200
-`;
+    redirects += `${route} /.netlify/functions/${id} 200\n`;
   }
 
-  if (apiRoutes.size > 0) {
-    redirects += `${basePath}/api/* /.netlify/functions/api-:splat 200`;
-  }
+  // if (apiRoutes.size > 0) {
+  //   redirects += `${basePath}/api/* /.netlify/functions/api-:splat 200`;
+  // }
 
   for (const [key, value] of apiRoutes.entries()) {
     const outputType = "api";
-    const { id, outputHref } = apiRoutes.get(key);
+    const { id, outputHref, route } = apiRoutes.get(key);
     const outputRoot = new URL(`./api/${id}/`, adapterOutputScratchUrl);
     const { assets = [] } = value;
 
@@ -159,8 +158,11 @@ async function netlifyAdapter(compilation) {
     // NOTE: All functions must live at the top level
     // https://github.com/netlify/netlify-lambda/issues/90#issuecomment-486047201
     await createOutputZip(id, outputType, outputRoot, projectDirectory);
+
+    redirects += `${route} /.netlify/functions/api-${id} 200\n`;
   }
 
+  console.log({ redirects, outputDir });
   if (redirects !== "") {
     await fs.writeFile(new URL("./_redirects", outputDir), redirects);
   }
