diff --git a/node_modules/@greenwood/cli/src/commands/build.js b/node_modules/@greenwood/cli/src/commands/build.js
index 6ea8d15..fba4ef7 100644
--- a/node_modules/@greenwood/cli/src/commands/build.js
+++ b/node_modules/@greenwood/cli/src/commands/build.js
@@ -71,6 +71,9 @@ const runProductionBuild = async (compilation) => {
       const prerenderPlugin = compilation.config.plugins.find(plugin => plugin.type === 'renderer')
         ? compilation.config.plugins.find(plugin => plugin.type === 'renderer').provider(compilation)
         : {};
+      const adapterPlugin = compilation.config.plugins.find(plugin => plugin.type === 'adapter')
+        ? compilation.config.plugins.find(plugin => plugin.type === 'adapter').provider(compilation)
+        : null;
 
       if (!await checkResourceExists(outputDir)) {
         await fs.mkdir(outputDir, {
@@ -114,6 +117,10 @@ const runProductionBuild = async (compilation) => {
       await bundleCompilation(compilation);
       await copyAssets(compilation);
 
+      if (adapterPlugin) {
+        await adapterPlugin();
+      }
+
       resolve();
     } catch (err) {
       reject(err);
diff --git a/node_modules/@greenwood/cli/src/lifecycles/config.js b/node_modules/@greenwood/cli/src/lifecycles/config.js
index f3afd0b..dd8bff4 100644
--- a/node_modules/@greenwood/cli/src/lifecycles/config.js
+++ b/node_modules/@greenwood/cli/src/lifecycles/config.js
@@ -35,7 +35,7 @@ const greenwoodPlugins = (await Promise.all([
 });
 
 const optimizations = ['default', 'none', 'static', 'inline'];
-const pluginTypes = ['copy', 'context', 'resource', 'rollup', 'server', 'source', 'renderer'];
+const pluginTypes = ['copy', 'context', 'resource', 'rollup', 'server', 'source', 'renderer', 'adapter'];
 const defaultConfig = {
   workspace: new URL('./src/', cwd),
   devServer: {
