diff --git a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
index 549448a..6c407f8 100644
--- a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
+++ b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
@@ -7,15 +7,26 @@ import {
 import { zip } from "zip-a-folder";
 
 // https://docs.netlify.com/functions/create/?fn-language=js
-function generateOutputFormat(id) {
+function generateOutputFormat(id, type, segmentKey) {
   const handlerAlias = "$handler";
-
+  // TODO: use `URLPattern` for extracting props after we upgrade to Node24
+  const extractProps = segmentKey
+    ? type === 'page'
+      ? `const props = { ${segmentKey}: rawUrl.split('/')[url.split('/').length - 2] }`
+      : `const props = { ${segmentKey}: rawUrl.split('?')[0].split('/').pop() }`
+    : 'const props = {}';
+  // const path = segment ? segment.pathname : route;
+  // export const config = {
+  //   path: ['${path}']
+  // };
   return `
     import { handler as ${handlerAlias} } from './${id}.js';
 
     export async function handler (event, context = {}) {
       const { rawUrl, body, headers = {}, httpMethod } = event;
       const contentType = headers['content-type'] || '';
+      ${extractProps}
+      console.log({ rawUrl, body, headers, httpMethod });
       let format = body;
 
       if (['GET', 'HEAD'].includes(httpMethod.toUpperCase())) {
@@ -41,7 +52,11 @@ function generateOutputFormat(id) {
         method: httpMethod,
         headers: new Headers(headers)
       });
-      const response = await ${handlerAlias}(request, context);
+      // TODO: what to do about context?
+      // const response = await ${handlerAlias}(request, context);
+      const response = '${type}' === 'page'
+        ? await ${handlerAlias}(request, props)
+        : await ${handlerAlias}(request, { props });
 
       return {
         statusCode: response.status,
@@ -52,10 +67,10 @@ function generateOutputFormat(id) {
   `;
 }
 
-async function setupOutputDirectory(id, outputRoot, outputType) {
+async function setupOutputDirectory(id, outputRoot, outputType, segmentKey) {
   const entryPoint = outputType === "api" ? id : `${id}.route`;
   const filename = outputType === "api" ? `api-${id}` : id;
-  const outputFormat = generateOutputFormat(entryPoint);
+  const outputFormat = generateOutputFormat(entryPoint, outputType, segmentKey);
 
   await fs.mkdir(outputRoot, { recursive: true });
   await fs.writeFile(new URL(`./${filename}.js`, outputRoot), outputFormat);
@@ -101,14 +116,14 @@ async function netlifyAdapter(compilation) {
   await fs.mkdir(adapterOutputUrl, { recursive: true });
 
   for (const page of ssrPages) {
-    const { id, outputHref, route } = page;
+    const { id, outputHref, route, segment } = page;
     const outputType = "page";
     const chunks = (await fs.readdir(outputDir)).filter(
       (file) => file.startsWith(`${id}.route.chunk`) && file.endsWith(".js"),
     );
     const outputRoot = new URL(`./${id}/`, adapterOutputScratchUrl);
 
-    await setupOutputDirectory(id, outputRoot, outputType);
+    await setupOutputDirectory(id, outputRoot, outputType, segment?.key);
 
     // handle user's actual route entry file
     await fs.cp(
@@ -131,17 +146,21 @@ async function netlifyAdapter(compilation) {
       projectDirectory,
     );
 
-    redirects += `${route} /.netlify/functions/${id} 200
+    console.log({ segment, route, id });
+    redirects += segment
+      ? `${segment.pathname.replace(`/:${segment.key}/`, `/:${segment.key}`)} /.netlify/functions/api-${id} 200
+`
+      : `${route} /.netlify/functions/${id} 200
 `;
   }
 
   for (const [key, value] of apiRoutes.entries()) {
     const outputType = "api";
-    const { id, outputHref, route } = apiRoutes.get(key);
+    const { id, outputHref, route, segment } = apiRoutes.get(key);
     const outputRoot = new URL(`./api/${id}/`, adapterOutputScratchUrl);
     const { assets = [] } = value;
 
-    await setupOutputDirectory(id, outputRoot, outputType);
+    await setupOutputDirectory(id, outputRoot, outputType, segment?.key);
 
     await fs.cp(new URL(outputHref), new URL(`./${id}.js`, outputRoot), { recursive: true });
 
@@ -155,11 +174,18 @@ async function netlifyAdapter(compilation) {
     // https://github.com/netlify/netlify-lambda/issues/90#issuecomment-486047201
     await createOutputZip(id, outputType, outputRoot, projectDirectory);
 
-    redirects += `${route} /.netlify/functions/api-${id} 200
+    console.log({ segment, route, id });
+    redirects += segment
+      ? `${segment.pathname} /.netlify/functions/api-${id} 200
+`
+      : `${route} /.netlify/functions/${id} 200
 `;
-  }
+}
 
+  // id: "/product/[id]",
+  // pattern: /^\/product\/([^/]+?)\/?$/,
   if (redirects !== "") {
+    console.log({ redirects, outputDir })
     await fs.writeFile(new URL("./_redirects", outputDir), redirects);
   }
 }
