diff --git a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
index 6fc9eef..ff6aea7 100644
--- a/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
+++ b/node_modules/@greenwood/plugin-adapter-netlify/src/index.js
@@ -18,6 +18,8 @@ function generateOutputFormat(id) {
       const contentType = headers['content-type'] || '';
       let format = body;
 
+      console.log({ event });
+      console.log({ body });
       if (['GET', 'HEAD'].includes(httpMethod.toUpperCase())) {
         format = null
       } else if (contentType.includes('application/x-www-form-urlencoded')) {
@@ -34,8 +36,6 @@ function generateOutputFormat(id) {
         // https://stackoverflow.com/a/43521052/417806
         format = formData;
         delete headers['content-type'];
-      } else if(contentType.includes('application/json')) {
-        format = JSON.stringify(body);
       }
 
       const request = new Request(rawUrl, {
